main/design_documents/function-multi-versioning.md

```diff
@@ -0,0 +1,66 @@
+# Function Multi Versioning support for Arm architectures
+
+Other architectures supports [function multi versioning]
+(https://gcc.gnu.org/wiki/FunctionMultiVersioning). This documents presents
+the rationale behind the design of the same feature for Arm.
+All toolchains and platforms should use the same source code formats to simplify
+the development and deployment of software and therefore increase the use of
+architectural features. Arm architectures are a bit more complex in view of
+feature availability therefore function versions would be better to be bound to
+features instead of cpu implementations. Also the kernels may limit the
+available features in the user space. Arm architectures are a bit more complex
+in view of feature availability therefore function versions would be better to
+be bound to features instead of cpu implementations. Also the kernels may limit
+the available features in the user space. Function Multi Versioning abbreviated
+to FMV.
+
+Why ACLE instead of ABI? FMV does not change how the function behaves from the
+ABI perspective. Because the caller does not know which version is actually
+called, the following holds for all the function derived from a function via
+FMV:
+
+1. the derived function have the same signature of the original function,
+2. the derived function obey to the same calling convention of the original
+function.
```

Comment:
Aren't rule 1 and 2 something that should actually go into the specs?

Reply:
"These attributes have no effect on the calling convention, all versions must
use the same calling convention." is in the spec.